### ISSUE TITLE
[JENKINS-39628] Handle skipped stages and parallels in the Pipeline Graph

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/blueocean-plugin#3404]
<div class="panel" style="border-width: 1px;"><div class="panelHeader" style="border-bottom-width: 1px;background-color: #8CC04F;color: white;"><b>Ready for release</b></div><div class="panelContent">
<p>This improvement is completed and scheduled for release in 1.2. Check the <a href="https://jenkins.io/projects/blueocean/roadmap/" class="external-link" target="_blank" rel="nofollow noopener">roadmap page</a> for updates. </p>
</div></div>

<p><b>Context</b><br/>
The Pipeline team have added the ability to "skip" stages in the Pipeline (<a href="https://issues.jenkins.io/browse/JENKINS-37781" title="Pipeline Config: ability to skip whole pipeline or stages based on boolean condition" class="issue-link" data-issue-key="JENKINS-37781"><del>JENKINS-37781</del></a>).</p>

<p><b>Scope</b></p>
<ul>
	<li>Skipped stages are shown as not built</li>
	<li>Declarative Pipeline will not provide flow nodes for skipped parallels so we can always assume that there won't be any skipped parallels.</li>
	<li>If the first stage is skipped then the pipeline should start from a "Start" node.</li>
	<li>If the last stage is skipped then the pipeline should end with a "End" node.</li>
	<li>Ensure that the editors start/end nodes match the style in this mockup</li>
</ul>

<p><b>Mockup</b><br/>
 <span class="image-wrap" style=""><a id="34840_thumb" href="https://issues.jenkins.io/secure/attachment/34840/34840_Skipped+Step.png" title="Skipped Step.png" file-preview-type="image" file-preview-id="34840" file-preview-title="Skipped Step.png"><img src="https://issues.jenkins.io/secure/thumbnail/34840/_thumb_34840.png" style="border: 0px solid black" role="presentation"/></a></span> </p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay">jamesdumay</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39628" target="_blank">Handle skipped stages and parallels in the Pipeline Graph</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=sophistifunk">sophistifunk</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Critical
<li><b>component(s)</b>: blueocean-plugin
<li><b>label(s)</b>: pipeline-improvement
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2017-07-31T21:43:40+00:00
<li><b>votes</b>: 10
<li><b>watchers</b>: 21
<li><b>imported</b>: 2025-12-01
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<div class="panel" style="border-width: 1px;"><div class="panelHeader" style="border-bottom-width: 1px;background-color: #8CC04F;color: white;"><b>Ready for release</b></div><div class="panelContent">
<p>This improvement is completed and scheduled for release in 1.2. Check the <a href="https://jenkins.io/projects/blueocean/roadmap/" class="external-link" target="_blank" rel="nofollow noopener">roadmap page</a> for updates. </p>
</div></div>

<p><b>Context</b>
The Pipeline team have added the ability to "skip" stages in the Pipeline (<a href="https://issues.jenkins.io/browse/JENKINS-37781" title="Pipeline Config: ability to skip whole pipeline or stages based on boolean condition" class="issue-link" data-issue-key="JENKINS-37781"><del>JENKINS-37781</del></a>).</p>

<p><b>Scope</b></p>
<ul>
	<li>Skipped stages are shown as not built</li>
	<li>Declarative Pipeline will not provide flow nodes for skipped parallels so we can always assume that there won't be any skipped parallels.</li>
	<li>If the first stage is skipped then the pipeline should start from a "Start" node.</li>
	<li>If the last stage is skipped then the pipeline should end with a "End" node.</li>
	<li>Ensure that the editors start/end nodes match the style in this mockup</li>
</ul>

<p><b>Mockup</b>
 <span class="image-wrap" style=""><a id="34840_thumb" href="https://issues.jenkins.io/secure/attachment/34840/34840_Skipped+Step.png" title="Skipped Step.png" file-preview-type="image" file-preview-id="34840" file-preview-title="Skipped Step.png"><img src="https://issues.jenkins.io/secure/thumbnail/34840/_thumb_34840.png" style="border: 0px solid black" role="presentation"/></a></span> </p></pre>
</details>
</details>
<details><summary><i>3 attachments</i></summary>

- [Blue Ocean Skipped Stage.PNG](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/40409/Blue%20Ocean%20Skipped%20Stage.PNG)
  > ![Blue Ocean Skipped Stage.PNG](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/40409/Blue%20Ocean%20Skipped%20Stage.PNG)
- [Skipped Step.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/34840/Skipped%20Step.png)
  > ![Skipped Step.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/34840/Skipped%20Step.png)
- [screenshot-1.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/34793/screenshot-1.png)
  > ![screenshot-1.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/34793/screenshot-1.png)
</details>

**Comment by jenkins-infra-bot** [comment]
- _Remote link associated with this issue:_


  - [Example Jenkinsfile](https://gist.github.com/abayer/02d13d1cf9546fc4b96179384a5f2a34)

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39628?focusedId=276657&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-276657">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=brody" class="user-hover" rel="brody">brody</a> we've got to come up with a way of demonstrating this.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=brody" class="user-hover" rel="brody">brody</a> we've got to come up with a way of demonstrating this.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/brody.png" title="brody's avatar" /> brody:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39628?focusedId=276850&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-276850">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a> – Initial idea</p>

<p> <span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/34793/34793_screenshot-1.png" style="border: 0px solid black" /></span></p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a> – Initial idea</p>

<p> <span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/34793/screenshot-1.png" style="border: 0px solid black" /></span></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/brody.png" title="brody's avatar" /> brody:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39628?focusedId=277056&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-277056">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Zeplin <a href="https://zpl.io/Z2m0Q0l" class="external-link" target="_blank" rel="nofollow noopener">https://zpl.io/Z2m0Q0l</a></p>

<p>Start / End nodes are added when First / Last stages are skipped. (see bottom of mockup)
Example of skipping 2 stages also at the bottom of the mockup</p>

<p> <span class="image-wrap" style=""><a id="34840_thumb" href="https://issues.jenkins.io/secure/attachment/34840/34840_Skipped+Step.png" title="Skipped Step.png" file-preview-type="image" file-preview-id="34840" file-preview-title="Skipped Step.png"><img src="https://issues.jenkins.io/secure/thumbnail/34840/_thumb_34840.png" style="border: 0px solid black" role="presentation"/></a></span> </p>
</pre>
</details>

</ul>
</details>

<p>Zeplin <a href="https://zpl.io/Z2m0Q0l" class="external-link" target="_blank" rel="nofollow noopener">https://zpl.io/Z2m0Q0l</a></p>

<p>Start / End nodes are added when First / Last stages are skipped. (see bottom of mockup)<br/>
Example of skipping 2 stages also at the bottom of the mockup</p>

<p> <span class="image-wrap" style=""><a id="34840_thumb" href="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/34840/Skipped%20Step.png" title="Skipped Step.png" file-preview-type="image" file-preview-id="34840" file-preview-title="Skipped Step.png"><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/34840/Skipped%20Step.png" style="border: 0px solid black" role="presentation"/></a></span> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>sophistifunk:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39628?focusedId=277696&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-277696">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Couple of questions:</p>

<ol>
	<li>Is there a compelling reason to only show start/end nodes when the first/last is skipped?</li>
	<li>Do we want to visually discriminate between "skipped because of new skip method" and "not run because previous stage failed"?</li>
	<li>Related to #2, let's say I have 10 parallel stages, but my infrastructure can only schedule 3 at a time, and one of the first 3 fails. Does Jenkins "skip" the other 7? If so, how do we want them to look?</li>
</ol>

</pre>
</details>

</ul>
</details>

<p>Couple of questions:</p>

<ol>
	<li>Is there a compelling reason to only show start/end nodes when the first/last is skipped?</li>
	<li>Do we want to visually discriminate between "skipped because of new skip method" and "not run because previous stage failed"?</li>
	<li>Related to #2, let's say I have 10 parallel stages, but my infrastructure can only schedule 3 at a time, and one of the first 3 fails. Does Jenkins "skip" the other 7? If so, how do we want them to look?</li>
</ol>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/brody.png" title="brody's avatar" /> brody:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39628?focusedId=277899&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-277899">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=sophistifunk" class="user-hover" rel="sophistifunk">sophistifunk</a>
1. I'm happy to them shown consistently, as long as we have a fallback if width starts becoming an issue.
2. IMO a stage that hasn't run, is a stage that hasn't run, regardless of it being skipped or not run because of previous failure.
3. If they're in "parallel" i'd run them all even if an earlier one fails, as the user would expect them all to be run</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=sophistifunk" class="user-hover" rel="sophistifunk">sophistifunk</a><br/>
1. I'm happy to them shown consistently, as long as we have a fallback if width starts becoming an issue.<br/>
2. IMO a stage that hasn't run, is a stage that hasn't run, regardless of it being skipped or not run because of previous failure.<br/>
3. If they're in "parallel" i'd run them all even if an earlier one fails, as the user would expect them all to be run</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39628?focusedId=290098&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-290098">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Bumped this up a bit - as a few people are asking for this as it is a most useful feature when they migrate to pipeline. This looks like a non trivial bit of work, so perhaps could be split so that a first pass just shows the node correctly as some skipped visual state before the svg code is changed - thoughts <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a></p>

</pre>
</details>

</ul>
</details>

<p>Bumped this up a bit - as a few people are asking for this as it is a most useful feature when they migrate to pipeline. This looks like a non trivial bit of work, so perhaps could be split so that a first pass just shows the node correctly as some skipped visual state before the svg code is changed - thoughts <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39628?focusedId=297127&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-297127">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=michaelneale" class="user-hover" rel="michaelneale">michaelneale</a> this is on the plan for the pipeline papercuts post-editor work.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=michaelneale" class="user-hover" rel="michaelneale">michaelneale</a> this is on the plan for the pipeline papercuts post-editor work.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>reinholdfuereder:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39628?focusedId=301533&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-301533">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Since I also just stumbled over the linked issue <a href="https://issues.jenkins.io/browse/JENKINS-44464" title="Blue Ocean shows 'Unknow' status for skipped Stages" class="issue-link" data-issue-key="JENKINS-44464"><del>JENKINS-44464</del></a> reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=wimnat" class="user-hover" rel="wimnat">wimnat</a> – (a) confusing/surprising colour and (b) "Unknow" typo in title, where ideally even the reason should be stated &#8211; maybe a more pragmatic (first?) approach could be:</p>
<ul>
	<li>replace purple with white colour (i.e. just a black circle with no/white filling)</li>
	<li>replace title message "Unknow" with "Skipped due to <span class="error">&#91;expression|branch|...&#93;</span>" (to give a reason, but presumably avoiding giving too many details)</li>
	<li>leave the stage connecting lines as is</li>
</ul>

</pre>
</details>

</ul>
</details>

<p>Since I also just stumbled over the linked issue <a href="https://issues.jenkins.io/browse/JENKINS-44464" title="Blue Ocean shows 'Unknow' status for skipped Stages" class="issue-link" data-issue-key="JENKINS-44464"><del>JENKINS-44464</del></a> reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=wimnat" class="user-hover" rel="wimnat">wimnat</a> – (a) confusing/surprising colour and (b) "Unknow" typo in title, where ideally even the reason should be stated &#8211; maybe a more pragmatic (first?) approach could be:</p>
<ul>
	<li>replace purple with white colour (i.e. just a black circle with no/white filling)</li>
	<li>replace title message "Unknow" with "Skipped due to <span class="error">&#91;expression|branch|...&#93;</span>" (to give a reason, but presumably avoiding giving too many details)</li>
	<li>leave the stage connecting lines as is</li>
</ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>sophistifunk:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39628?focusedId=301596&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-301596">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Purple isn't "skipped", it's "unknown status" - it's meant to stand out because it means either the UI needs to be updated or the server is sending bad data.</p>
</pre>
</details>

</ul>
</details>

<p>Purple isn't "skipped", it's "unknown status" - it's meant to stand out because it means either the UI needs to be updated or the server is sending bad data.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/wimnat.svg" title="wimnat's avatar" /> wimnat:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39628?focusedId=301638&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-301638">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p> I would like to be able to completely hide skipped stages.</p>

<p>I have a pipeline that does a different build depending on branch - i don't really want to see something like this:</p>

<p>Checkout -&gt; Build (develop) (SKIPPED) -&gt; Build (master) -&gt; Build (feature) (SKIPPED) -&gt; Deploy</p>

<p>It would be great if there was a per project toggle that would hide the skipped steps.  With this toggle on it would look something like:</p>

<p>Checkout -&gt; Build (master) -&gt; Deploy</p>
</pre>
</details>

</ul>
</details>

<p> I would like to be able to completely hide skipped stages.</p>

<p>I have a pipeline that does a different build depending on branch - i don't really want to see something like this:</p>

<p>Checkout -> Build (develop) (SKIPPED) -> Build (master) -> Build (feature) (SKIPPED) -> Deploy</p>

<p>It would be great if there was a per project toggle that would hide the skipped steps.  With this toggle on it would look something like:</p>

<p>Checkout -> Build (master) -> Deploy</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>thuybdojenkins:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39628?focusedId=320001&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-320001">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I second Mr. White.  In my work, I have a pipeline that contains more than 20 stages.  Some can be executed in parallel and some have to be sequential.  Each stage is executed based on a condition.  I would like to hide the stage that is not executed.  See below screenshot. </p>

<p>  <span class="image-wrap" style=""><a id="40409_thumb" href="https://issues.jenkins.io/secure/attachment/40409/40409_Blue+Ocean+Skipped+Stage.PNG" title="Blue Ocean Skipped Stage.PNG" file-preview-type="image" file-preview-id="40409" file-preview-title="Blue Ocean Skipped Stage.PNG"><img src="https://issues.jenkins.io/secure/thumbnail/40409/_thumb_40409.png" style="border: 0px solid black" role="presentation"/></a></span></p>

<p>For this pipeline, I have to scroll to the right often to see the status of the stage.  It would be nice if the skipped stages either collapsed into "skipped stages" to not display at all.  I prefer not to display them at all.  Collapsing the skipped stages may not work if every other stage is skipped.</p>

<p> </p>

<p> </p>
</pre>
</details>

</ul>
</details>

<p>I second Mr. White.  In my work, I have a pipeline that contains more than 20 stages.  Some can be executed in parallel and some have to be sequential.  Each stage is executed based on a condition.  I would like to hide the stage that is not executed.  See below screenshot. </p>

<p>  <span class="image-wrap" style=""><a id="40409_thumb" href="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/40409/Blue%20Ocean%20Skipped%20Stage.PNG" title="Blue Ocean Skipped Stage.PNG" file-preview-type="image" file-preview-id="40409" file-preview-title="Blue Ocean Skipped Stage.PNG"><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/40409/Blue%20Ocean%20Skipped%20Stage.PNG" style="border: 0px solid black" role="presentation"/></a></span></p>

<p>For this pipeline, I have to scroll to the right often to see the status of the stage.  It would be nice if the skipped stages either collapsed into "skipped stages" to not display at all.  I prefer not to display them at all.  Collapsing the skipped stages may not work if every other stage is skipped.</p>

<p> </p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39628?focusedId=320015&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-320015">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=thuybdojenkins" class="user-hover" rel="thuybdojenkins">thuybdojenkins</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=wimnat" class="user-hover" rel="wimnat">wimnat</a> could I see a sanitised sample of your pipelines if possible? </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=thuybdojenkins" class="user-hover" rel="thuybdojenkins">thuybdojenkins</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=wimnat" class="user-hover" rel="wimnat">wimnat</a> could I see a sanitised sample of your pipelines if possible? </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39628?focusedId=320021&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-320021">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=thuybdojenkins" class="user-hover" rel="thuybdojenkins">thuybdojenkins</a> could you please open up a new feature request for skipping with an example Pipeline? Reiterating the use case you have here would be good (generally we do not accept feature requests on already resolved tickets for other features)</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=thuybdojenkins" class="user-hover" rel="thuybdojenkins">thuybdojenkins</a> could you please open up a new feature request for skipping with an example Pipeline? Reiterating the use case you have here would be good (generally we do not accept feature requests on already resolved tickets for other features)</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>thuybdojenkins:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39628?focusedId=320109&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-320109">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=michaelneale" class="user-hover" rel="michaelneale">michaelneale</a> I attached the pipeline view in my comment.  Do you need the groovy script I use?  If so, I will include in the new feature request as suggested by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a>.</p>

<p>Here is a snippet of a stage in my pipeline.  Other stages are similar.</p>

<p>stage("dps") {</p>

<p>    when {
        expression { return params.enableDps }
    }</p>

<p>    steps {
        timeout(time: 10, unit: 'HOURS') {
            ecdmItestRun(
                vmIp : "${params.vmIp}",
                itestPath : "itests/restapi/dps",
                itestName : "DPS",
                additionalArgs : "-Prun-its,run-its-irvine"
                )
        }
    }
 }</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=michaelneale" class="user-hover" rel="michaelneale">michaelneale</a> I attached the pipeline view in my comment.  Do you need the groovy script I use?  If so, I will include in the new feature request as suggested by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a>.</p>

<p>Here is a snippet of a stage in my pipeline.  Other stages are similar.</p>

<p>stage("dps") {</p>

<p>    when {<br/>
        expression { return params.enableDps }<br/>
    }</p>

<p>    steps {<br/>
        timeout(time: 10, unit: 'HOURS') {<br/>
            ecdmItestRun(<br/>
                vmIp : "${params.vmIp}",<br/>
                itestPath : "itests/restapi/dps",<br/>
                itestName : "DPS",<br/>
                additionalArgs : "-Prun-its,run-its-irvine"<br/>
                )<br/>
        }<br/>
    }<br/>
 }</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ruudp.svg" title="ruudp's avatar" /> ruudp:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39628?focusedId=321194&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-321194">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Is this working also for Parallel stages on different nodes?  I am facing at this moment some strange behavior,</p>

<p>if  one of the when statements is false:</p>

<p>Blue ocean shows "Queued Waiting for next available executor on ....." until the end of the complete parallel execution,  but i see in the console everything is running</p>

<p>If all when statements are true i dont see anny issue</p>

<p>below a small example code</p>

<p> parallel {</p>

<p>        stage(A') {
          when {expression {params.A}}
            steps {node(label: 'A') {bla bla }}}
        stage('Bt') {
          when {expression {params.B}}
            steps {node(label: 'B') {bla bla}}}
        stage('C') {
          when {expression {params.C}}
            steps {node(label: 'C') {bla bla}}}</p>

<p>}</p>

<p> </p>
</pre>
</details>

</ul>
</details>

<p>Is this working also for Parallel stages on different nodes?  I am facing at this moment some strange behavior,</p>

<p>if  one of the when statements is false:</p>

<p>Blue ocean shows "Queued Waiting for next available executor on ....." until the end of the complete parallel execution,  but i see in the console everything is running</p>

<p>If all when statements are true i dont see anny issue</p>

<p>below a small example code</p>

<p> parallel {</p>

<p>        stage(A') {<br/>
          when {expression {params.A}}<br/>
            steps {node(label: 'A') {bla bla }}}<br/>
        stage('Bt') {<br/>
          when {expression {params.B}}<br/>
            steps {node(label: 'B') {bla bla}}}<br/>
        stage('C') {<br/>
          when {expression {params.C}}<br/>
            steps {node(label: 'C') {bla bla}}}</p>

<p>}</p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39628?focusedId=325394&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-325394">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ruudp" class="user-hover" rel="ruudp">ruudp</a> I don't think it works in parallel yet (feel free to open a new ticket though)</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ruudp" class="user-hover" rel="ruudp">ruudp</a> I don't think it works in parallel yet (feel free to open a new ticket though)</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>comscience15:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39628?focusedId=329486&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-329486">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=sophistifunk" class="user-hover" rel="sophistifunk">sophistifunk</a> Hi guys, I wonder instead of adding "when{....}" to determine whether the stage will be skipped or not, can we have user input feature to ask if they want to skip the stage or not. Do we have this feature too?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=sophistifunk" class="user-hover" rel="sophistifunk">sophistifunk</a> Hi guys, I wonder instead of adding "when{....}" to determine whether the stage will be skipped or not, can we have user input feature to ask if they want to skip the stage or not. Do we have this feature too?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>sophistifunk:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39628?focusedId=329487&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-329487">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I imagine you'd achieve that easily enough with an input step to set a variable?</p>
</pre>
</details>

</ul>
</details>

<p>I imagine you'd achieve that easily enough with an input step to set a variable?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39628?focusedId=329489&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-329489">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=comscience15" class="user-hover" rel="comscience15">comscience15</a> not explicitly no - but like Josh says, could have an input step in an earlier stage that captures some environment variable that is then referred to in 'when'. A parametrised job could also provide the same variable I think. Can't put an input inside a 'when' (I think). </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=comscience15" class="user-hover" rel="comscience15">comscience15</a> not explicitly no - but like Josh says, could have an input step in an earlier stage that captures some environment variable that is then referred to in 'when'. A parametrised job could also provide the same variable I think. Can't put an input inside a 'when' (I think). </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ruudp.svg" title="ruudp's avatar" /> ruudp:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39628?focusedId=329726&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-329726">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>it is fixed with an update of jenkins, at this moment this works fine for me</p>
</pre>
</details>

</ul>
</details>

<p>it is fixed with an update of jenkins, at this moment this works fine for me</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>comscience15:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39628?focusedId=329832&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-329832">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=sophistifunk" class="user-hover" rel="sophistifunk">sophistifunk</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=michaelneale" class="user-hover" rel="michaelneale">michaelneale</a> Thanks for your feedback. Yes, it is easy to set a variable. My thought is just about if there is a complicated case which is sometimes is outside the set up variable. So if we can get it with user input too, it would be great. </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=sophistifunk" class="user-hover" rel="sophistifunk">sophistifunk</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=michaelneale" class="user-hover" rel="michaelneale">michaelneale</a> Thanks for your feedback. Yes, it is easy to set a variable. My thought is just about if there is a complicated case which is sometimes is outside the set up variable. So if we can get it with user input too, it would be great. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>lhupfeldt:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39628?focusedId=412472&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-412472">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The graph in <b>scripted</b> pipelines does not show the line flowing around the skipped stage, but through it, like executed steps.</p>

<p>The steps are shown as "No steps" and the <b>full job</b> log is displayed.</p>
</pre>
</details>

</ul>
</details>

<p>The graph in <b>scripted</b> pipelines does not show the line flowing around the skipped stage, but through it, like executed steps.</p>

<p>The steps are shown as "No steps" and the <b>full job</b> log is displayed.</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `is_duplicated_by` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-44464%22">JENKINS-44464</a>]</i>
